File: src/content/blog/porting-htomd-beyond-python/snippets/conditional-clutter.py
def conditional_clutter(node: Node, stat: Stats) -> bool:
    if node.tag in {"#document", "html", "body", "main", "article"}:
        return False
    tokens = hints(node)
    if tokens & REFERENCES or node.attrs.get("role") == "note":
        return False
    if not tokens & NEGATIVE or node.tag in EVIDENCE | HEADINGS:
        return False
    # Class hints alone never delete a node: require structural corroboration.
    if stat.density > 0.35 or stat.controls:
        return True
    if "footer" in tokens and stat.density > 0.15:
        return True
    if node.tag == "aside" and not stat.code and not stat.cells:
        return True
    if stat.blocks == 0 and stat.characters < 180:
        return True
    if tokens & {"comments", "comment"}:
        children = elements(node)
        return sum(bool(hints(child) & {"comment", "reply"}) for child in children) >= 2
    return False

File: src/content/blog/porting-htomd-beyond-python/snippets/conditional-clutter.ts
function conditionalClutter(node: Node, stat: Stats): boolean {
  if (["#document", "html", "body", "main", "article"].includes(node.tag)) {
    return false;
  }
  const tokens = hints(node);
  if (intersects(tokens, REFERENCES) || node.attrs.get("role") === "note") {
    return false;
  }
  if (!intersects(tokens, NEGATIVE) || EVIDENCE.has(node.tag) || HEADINGS.has(node.tag)) {
    return false;
  }
  if (stat.density > 0.35 || stat.controls) {
    return true;
  }
  if (tokens.has("footer") && stat.density > 0.15) {
    return true;
  }
  if (node.tag === "aside" && !stat.code && !stat.cells) {
    return true;
  }
  if (!stat.blocks && stat.characters < 180) {
    return true;
  }
  if (tokens.has("comments") || tokens.has("comment")) {
    return (
      elements(node).filter((child) => {
        const childHints = hints(child);
        return childHints.has("comment") || childHints.has("reply");
      }).length >= 2
    );
  }
  return false;
}
